test/src/unit-cppapi-float-scaling-filter.cc

```diff
@@ -0,0 +1,218 @@
+    T val_float = static_cast<T>(scale * static_cast<T>(val) + offset);
+    expected_a.push_back(val_float);
+  }
+
+  tiledb_layout_t layout_type =
+      array_type == TILEDB_SPARSE ? TILEDB_UNORDERED : TILEDB_ROW_MAJOR;
+
+  Array array_w(ctx, array_name, TILEDB_WRITE);
+  Query query_w(ctx, array_w);
+  query_w.set_layout(layout_type).set_data_buffer("a", a_write);
+
+  if (array_type == TILEDB_SPARSE) {
+    query_w.set_data_buffer("rows", row_dims).set_data_buffer("cols", col_dims);
+  }
+
+  query_w.submit();
+  query_w.finalize();
+  array_w.close();
+
+  // Open and read the entire array.
+  std::vector<T> a_data_read(dim_hi * dim_hi, 0.0);
+  Array array_r(ctx, array_name, TILEDB_READ);
+  Query query_r(ctx, array_r);
+  query_r.set_layout(TILEDB_ROW_MAJOR).set_data_buffer("a", a_data_read);
+
+  if (array_type == TILEDB_DENSE) {
+    int range[] = {1, dim_hi};
+    query_r.add_range("rows", range[0], range[1])
+        .add_range("cols", range[0], range[1]);
+  }
+
+  query_r.submit();
+
+  // Check for results.
+  size_t total_num_elements = static_cast<size_t>(dim_hi * dim_hi);
+  auto table = query_r.result_buffer_elements();
+  REQUIRE(table.size() == 1);
+  REQUIRE(table["a"].first == 0);
+  REQUIRE(table["a"].second == total_num_elements);
+
+  for (size_t i = 0; i < total_num_elements; ++i) {
+    CHECK(
+        fabs(a_data_read[i] - expected_a[i]) <
+        std::numeric_limits<T>::epsilon());
+  }
+
+  query_r.finalize();
+  array_r.close();
+}
+
+TEST_CASE(
+    "C++ API: Float Scaling Filter list on array",
+    "[cppapi][filter][float-scaling]") {
+  // Setup.
+  Context ctx;
+  VFS vfs(ctx);
+
+  if (vfs.is_dir(array_name))
+    vfs.remove_dir(array_name);
+
+  auto array_type = GENERATE(TILEDB_SPARSE, TILEDB_DENSE);
+
+  SECTION("Testing FLOAT32 attribute type, bit_width = 1.") {
+    float_scaling_filter_api_test<float, int8_t>(ctx, array_type);
+  }
+
+  SECTION("Testing FLOAT32 attribute type, bit_width = 2.") {
+    float_scaling_filter_api_test<float, int16_t>(ctx, array_type);
+  }
+
+  SECTION("Testing FLOAT32 attribute type, bit_width = 4.") {
+    float_scaling_filter_api_test<float, int32_t>(ctx, array_type);
+  }
+
+  SECTION("Testing FLOAT32 attribute type, bit_width = 8.") {
+    float_scaling_filter_api_test<float, int64_t>(ctx, array_type);
+  }
+
+  SECTION("Testing FLOAT64 attribute type, bit_width = 1.") {
+    float_scaling_filter_api_test<double, int8_t>(ctx, array_type);
+  }
+
+  SECTION("Testing FLOAT64 attribute type, bit_width = 2.") {
+    float_scaling_filter_api_test<double, int16_t>(ctx, array_type);
+  }
+
+  SECTION("Testing FLOAT64 attribute type, bit_width = 4.") {
+    float_scaling_filter_api_test<double, int32_t>(ctx, array_type);
+  }
+
+  SECTION("Testing FLOAT64 attribute type, bit_width = 8.") {
+    float_scaling_filter_api_test<double, int64_t>(ctx, array_type);
+  }
+
+  // Teardown.
+  if (vfs.is_dir(array_name))
+    vfs.remove_dir(array_name);
+}
```

Comment:
Optional: This can be simplified using TEMPLATE_TEST_CASE or TEMPLATE_PRODUCT_TEST_CASE. See [Catch2 docs](https://github.com/catchorg/Catch2/blob/devel/docs/test-cases-and-sections.md#type-parametrised-test-cases).